### DISCUSSION TITLE
Nextjs converts "typeof window" -> "object" in web workers?

### DISCUSSION THREAD

**Discussion by m11m** [vercel/next.js#39605] [Help]
Nextjs is converting "typeof window" to "object" in web workers, is this intentional?

From my understanding, Nextjs converts "typeof window" to "object" for the client, and to "undefined" for the server, to help with code splitting / purging. But it also converts it to "object" in web workers, where window should be "undefined".

This causes errors when you try to run some libraries, such as [three.js](https://github.com/mrdoob/three.js/blob/9f4de99828c05e71c47e6de0beb4c6e7652e486a/src/Three.js#L168), in a web worker. It uses a "typeof window" check to see if its in a worker or not, but Nextjs is breaking this check.

Whether this is a bug or just a quirk of Nextjs, does anyone know of a way to get around this? I'm trying to run threejs in a worker.

**Comment by m11m** [comment]
so it looks like it doesnt mess up `typeof window` checks in the the worker file itself, but any `typeof window` checks from imported modules do get converted to `object`, which is incorrect.

**Comment by icyJoseph** [comment]
I think this is a quirk of SWC/Babel, who eagerly replace this things at compile time, not necessarily a Next.js issue. 

I found this thread that is highly relevant, I believe anyway, https://github.com/vercel/next.js/issues/36514#issuecomment-1112105991 - 

And here's the documented SWC behavior, https://swc.rs/docs/configuration/compilation#jsctransformoptimizerglobals. 

Still not sure on how to proceed though... somehow we should tell SWC, that it shouldn't compile the worker code I guess?

**Comment by noe** [comment]
This affects emscripten-based code, which uses `typeof window === 'object'`  to check if they are running in a browser. They subsequently access `window`, leading to a ReferenceError.
